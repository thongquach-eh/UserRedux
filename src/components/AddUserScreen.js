// @flow
import React, {useLayoutEffect, useState, useCallback} from 'react';
import type {Node} from 'react';
import {View, Text, Button, StyleSheet, Image, ScrollView} from 'react-native';
import type {RouteProp} from '@react-navigation/native';
import type {StackNavigationProp} from '@react-navigation/stack';
import {useDispatch, useSelector} from 'react-redux';
import {addUser} from '../UsersAction';
import _ from 'lodash';
import type {UserDispatch, User, UserState} from '../types.js';
import UserInput from './UserInput';
import DateTimePicker from '@react-native-community/datetimepicker';
import {Picker} from '@react-native-picker/picker';
import {launchImageLibrary} from 'react-native-image-picker';

type AddUserScreenProps = {
  navigation: StackNavigationProp<any, any>,
  route: RouteProp<any, any>,
};

const styles = StyleSheet.create({
  avatar: {
    width: 150,
    height: 150,
    alignSelf: 'center',
  },
  detailsContainer: {
    margin: 10,
  },
  locationLabel: {
    fontSize: 18,
    marginVertical: 10,
  },
  locationContainer: {
    marginLeft: 10,
  },
  inputContainer: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  inputLabel: {
    flexBasis: '25%',
    fontSize: 16,
    marginVertical: 3,
  },
  inputValue: {
    flexBasis: '75%',
    fontSize: 16,
    marginVertical: 3,
  },
});

const AddUserScreen = ({navigation, route}: AddUserScreenProps): Node => {
  const dispatch = useDispatch<UserDispatch>();
  const users = useSelector((state: UserState) => state.users);
  const newUser = {
    name: {
      first: '',
      last: '',
    },
    email: '',
    phone: '',
    cell: '',
    gender: '',
  };
  const [user, setUser] = useState<User>(newUser);

  const saveUser = useCallback(() => {
    for (var otherUser of users) {
      if (otherUser.email === user.email) {
        // eslint-disable-next-line no-alert
        alert('This email address has already been used.');
        return;
      }
    }
    dispatch(addUser(user));
    navigation.goBack();
  }, [dispatch, navigation, user, users]);

  useLayoutEffect(() => {
    navigation.setOptions({
      headerRight: () => <Button title="Save" onPress={() => saveUser()} />,
    });
  }, [navigation, saveUser]);

  const writeValue = useCallback(
    (value: string | number, path: string): void => {
      const updatedUser = _.clone(_.set(user, path, value));
      setUser(updatedUser);
    },
    [user],
  );

  const handleChooseAvatar = useCallback(() => {
    const options = {
      noData: true,
    };
    launchImageLibrary(options, response => {
      if (response.uri) {
        writeValue(response.uri, 'picture.large');
        writeValue(response.uri, 'picture.medium');
        writeValue(response.uri, 'picture.thumbnail');
      }
    });
  }, [writeValue]);

  return (
    <ScrollView style={styles.detailsContainer}>
      <Image source={{uri: user.picture?.large}} style={styles.avatar} />
      <Button title="Choose Avatar" onPress={handleChooseAvatar} />
      <View style={styles.inputContainer}>
        <Text style={styles.inputLabel}>Title:</Text>
        <Picker
          style={styles.inputValue}
          selectedValue={user.name.title}
          onValueChange={(val: string | number) =>
            writeValue(val, 'name.title')
          }>
          <Picker.Item label="Mr" value="Mr" />
          <Picker.Item label="Ms" value="Ms" />
          <Picker.Item label="Mrs" value="Mrs" />
        </Picker>
      </View>
      <UserInput
        label="First name:"
        onChangeText={(val: string) => writeValue(val, 'name.first')}
      />
      <UserInput
        label="Last name:"
        onChangeText={(val: string) => writeValue(val, 'name.last')}
      />
      <View style={styles.inputContainer}>
        <Text style={styles.inputLabel}>Gender:</Text>
        <Picker
          style={styles.inputValue}
          selectedValue={user.gender}
          onValueChange={(val: string | number) => writeValue(val, 'gender')}>
          <Picker.Item label="male" value="male" />
          <Picker.Item label="female" value="female" />
        </Picker>
      </View>
      <View style={styles.inputContainer}>
        <Text style={styles.inputLabel}>Birth date:</Text>
        <DateTimePicker
          value={user.dob?.date ? new Date(user.dob.date) : new Date()}
          onChange={(event, val: string) =>
            writeValue(new Date(val).toISOString(), 'dob.date')
          }
          style={styles.inputValue}
        />
      </View>
      <UserInput
        label="Email:"
        onChangeText={(val: string) => writeValue(val, 'email')}
      />
      <UserInput
        label="Phone:"
        onChangeText={(val: string) => writeValue(val, 'phone')}
      />
      <UserInput
        label="Cellphone:"
        onChangeText={(val: string) => writeValue(val, 'cell')}
      />
      <Text style={styles.locationLabel}>Location:</Text>
      <View style={styles.locationContainer}>
        <UserInput
          label="Number"
          onChangeText={(val: string) =>
            writeValue(parseInt(val, 10), 'location.street.number')
          }
        />
        <UserInput
          label="Street"
          onChangeText={(val: string) =>
            writeValue(val, 'location.street.name')
          }
        />
        <UserInput
          label="City"
          onChangeText={(val: string) => writeValue(val, 'location.city')}
        />
        <UserInput
          label="State"
          onChangeText={(val: string) => writeValue(val, 'location.state')}
        />
        <UserInput
          label="Postcode"
          onChangeText={(val: string) => writeValue(val, 'location.postcode')}
        />
        <UserInput
          label="Country"
          onChangeText={(val: string) => writeValue(val, 'location.country')}
        />
      </View>
    </ScrollView>
  );
};

export default AddUserScreen;
