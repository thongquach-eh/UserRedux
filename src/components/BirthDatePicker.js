// @flow
import * as React from 'react';
import {StyleSheet, Text, View} from 'react-native';
import DateTimePicker from '@react-native-community/datetimepicker';
import {Field} from 'redux-form';
import type {FieldProps} from 'redux-form';

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    flexWrap: 'wrap',
    alignItems: 'center',
  },
  label: {
    flexBasis: '25%',
    fontSize: 16,
    marginVertical: 3,
  },
  value: {
    flexBasis: '75%',
    fontSize: 16,
    marginVertical: 3,
  },
});

const renderDatePicker = (props: FieldProps) => {
  const {
    input: {onChange, value},
  } = props;

  const onDateChange = (event, val) => {
    onChange(val.toISOString());
  };

  return (
    <DateTimePicker
      value={value ? new Date(value) : new Date()}
      style={styles.value}
      onChange={(event, val) => onDateChange(event, val)}
    />
  );
};

type BirthDatePickerProps = {
  label: string,
};

const BirthDatePicker = (props: BirthDatePickerProps): React.Node => {
  const {label} = props;

  return (
    <View style={styles.container}>
      <Text style={styles.label}>{label}</Text>
      <Field name="dob.date" component={renderDatePicker} />
    </View>
  );
};

export default BirthDatePicker;
