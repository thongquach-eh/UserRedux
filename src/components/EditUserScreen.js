// @flow
import React, {useLayoutEffect, useState, useCallback} from 'react';
import type {Node} from 'react';
import {
  View,
  Text,
  Button,
  StyleSheet,
  Image,
  ScrollView,
  KeyboardAvoidingView,
  Platform,
} from 'react-native';
import type {RouteProp} from '@react-navigation/native';
import type {StackNavigationProp} from '@react-navigation/stack';
import {useHeaderHeight} from '@react-navigation/stack';
import {useDispatch, useSelector} from 'react-redux';
import {editUser} from '../UsersAction';
import {set} from 'lodash/fp';
import type {
  UserDispatch,
  User,
  UserState,
  RootStackParamList,
} from '../types.js';
import UserInput from './UserInput';
import DateTimePicker from '@react-native-community/datetimepicker';
import {Picker} from '@react-native-picker/picker';
import {launchImageLibrary} from 'react-native-image-picker';
import {validateUser} from './Validations';

type EditUserScreenProps = {
  navigation: StackNavigationProp<RootStackParamList, 'EditUser'>,
  route: RouteProp<RootStackParamList, 'EditUser'>,
};

type EditUserPanelProps = {
  ...EditUserScreenProps,
  user: User,
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  avatar: {
    width: 150,
    height: 150,
    alignSelf: 'center',
  },
  detailsContainer: {
    margin: 10,
  },
  locationLabel: {
    fontSize: 18,
    marginVertical: 10,
  },
  locationContainer: {
    marginLeft: 10,
  },
  inputContainer: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  inputLabel: {
    flexBasis: '25%',
    fontSize: 16,
    marginVertical: 3,
  },
  inputValue: {
    flexBasis: '75%',
    fontSize: 16,
    marginVertical: 3,
  },
  pickerItem: {
    fontSize: 16,
  },
});

const EditUserPanel = ({user, navigation, route}: EditUserPanelProps): Node => {
  const dispatch = useDispatch<UserDispatch>();
  const users = useSelector((state: UserState) => state.users);
  const [draftUser, setDraftUser] = useState<User>(user);

  const saveEditedUser = useCallback(
    u => {
      if (validateUser(u, users)) {
        dispatch(editUser(u));
        return true;
      }
      return false;
    },
    [dispatch, users],
  );

  useLayoutEffect(() => {
    navigation.setOptions({
      headerRight: () => (
        <Button
          title="Save"
          onPress={() => {
            if (saveEditedUser(draftUser)) {
              navigation.goBack();
            }
          }}
        />
      ),
    });
  }, [saveEditedUser, navigation, draftUser]);

  const writeValue = useCallback(
    (value: any, path: string, u: User) => {
      const updatedUser = set(path, value)(u);
      setDraftUser(updatedUser);
    },
    [setDraftUser],
  );

  const handleChooseAvatar = useCallback(
    (u: User) => {
      const options = {
        noData: true,
      };
      launchImageLibrary(options, response => {
        if (response.uri) {
          const picture = {
            large: response.uri,
            medium: response.uri,
            thumbnail: response.uri,
          };
          writeValue(picture, 'picture', u);
        }
      });
    },
    [writeValue],
  );

  return (
    <KeyboardAvoidingView
      behavior={Platform.OS === 'ios' ? 'padding' : null}
      style={styles.container}
      keyboardVerticalOffset={useHeaderHeight() + 20}>
      <ScrollView style={styles.detailsContainer}>
        <Image source={{uri: draftUser.picture?.large}} style={styles.avatar} />
        <Button
          title="Choose Avatar"
          onPress={() => handleChooseAvatar(draftUser)}
        />
        <View style={styles.inputContainer}>
          <Text style={styles.inputLabel}>Title:</Text>
          <Picker
            style={styles.inputValue}
            itemStyle={styles.pickerItem}
            selectedValue={draftUser.name.title}
            onValueChange={(val: string | number) =>
              writeValue(val, 'name.title', draftUser)
            }>
            <Picker.Item label="Mr" value="Mr" />
            <Picker.Item label="Ms" value="Ms" />
            <Picker.Item label="Mrs" value="Mrs" />
          </Picker>
        </View>
        <UserInput
          label="First name:"
          value={draftUser.name.first}
          onChangeText={(val: string) =>
            writeValue(val, 'name.first', draftUser)
          }
        />
        <UserInput
          label="Last name:"
          value={draftUser.name.last}
          onChangeText={(val: string) =>
            writeValue(val, 'name.last', draftUser)
          }
        />
        <View style={styles.inputContainer}>
          <Text style={styles.inputLabel}>Gender:</Text>
          <Picker
            style={styles.inputValue}
            itemStyle={styles.pickerItem}
            selectedValue={draftUser.gender}
            onValueChange={(val: string | number) =>
              writeValue(val, 'gender', draftUser)
            }>
            <Picker.Item label="male" value="male" />
            <Picker.Item label="female" value="female" />
          </Picker>
        </View>
        <View style={styles.inputContainer}>
          <Text style={styles.inputLabel}>Birth date:</Text>
          <DateTimePicker
            value={
              draftUser.dob?.date ? new Date(draftUser.dob.date) : new Date()
            }
            onChange={(event, val: string) =>
              writeValue(new Date(val).toISOString(), 'dob.date', draftUser)
            }
            style={styles.inputValue}
          />
        </View>
        <UserInput
          label="Email:"
          value={draftUser.email}
          onChangeText={(val: string) => writeValue(val, 'email', draftUser)}
        />
        <UserInput
          label="Phone:"
          value={draftUser.phone}
          onChangeText={(val: string) => writeValue(val, 'phone', draftUser)}
        />
        <UserInput
          label="Cellphone:"
          value={draftUser.cell}
          onChangeText={(val: string) => writeValue(val, 'cell', draftUser)}
        />
        <Text style={styles.locationLabel}>Location:</Text>
        <View style={styles.locationContainer}>
          <UserInput
            label="Number"
            value={draftUser.location?.street.number?.toString()}
            onChangeText={(val: string) =>
              writeValue(parseInt(val, 10), 'location.street.number', draftUser)
            }
          />
          <UserInput
            label="Street"
            value={draftUser.location?.street.name}
            onChangeText={(val: string) =>
              writeValue(val, 'location.street.name', draftUser)
            }
          />
          <UserInput
            label="City"
            value={draftUser.location?.city}
            onChangeText={(val: string) =>
              writeValue(val, 'location.city', draftUser)
            }
          />
          <UserInput
            label="State"
            value={draftUser.location?.state}
            onChangeText={(val: string) =>
              writeValue(val, 'location.state', draftUser)
            }
          />
          <UserInput
            label="Postcode"
            value={draftUser.location?.postcode.toString()}
            onChangeText={(val: string) =>
              writeValue(val, 'location.postcode', draftUser)
            }
          />
          <UserInput
            label="Country"
            value={draftUser.location?.country}
            onChangeText={(val: string) =>
              writeValue(val, 'location.country', draftUser)
            }
          />
        </View>
      </ScrollView>
    </KeyboardAvoidingView>
  );
};

const EditUserScreen = ({navigation, route}: EditUserScreenProps): Node => {
  const {id} = route.params;
  const retrievedUser: ?User = useSelector((state: UserState) =>
    state.users.find(u => u.id === id),
  );

  if (retrievedUser == null) {
    return <Text>Error when trying to display the user with email: {id}!</Text>;
  } else {
    return (
      <EditUserPanel
        user={retrievedUser}
        navigation={navigation}
        route={route}
      />
    );
  }
};

export default EditUserScreen;
