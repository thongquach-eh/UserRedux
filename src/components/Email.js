// @flow
import * as React from 'react';
import {Text, StyleSheet, Linking} from 'react-native';
import type {TextStyleProp} from 'react-native/Libraries/StyleSheet/StyleSheet';

const styles = StyleSheet.create({
  default: {
    color: 'blue',
    textDecorationLine: 'underline',
  },
});

const Email = ({
  value,
  style,
}: {
  value: string,
  style?: TextStyleProp,
}): React.Node => (
  <Text
    onPress={() => Linking.openURL('mailto:' + value)}
    style={[styles.default, style]}>
    {value}
  </Text>
);

export default Email;
