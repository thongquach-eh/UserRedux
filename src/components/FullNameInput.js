// @flow
import * as React from 'react';
import {StyleSheet, View} from 'react-native';
import StringInput from './StringInput';
import TitlePicker from './TitlePicker';
import type {FullName} from '../types.js';

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    flexWrap: 'wrap',
  },
});

const FullNameInput = ({
  name,
  onChange,
}: {
  name: FullName,
  onChange: (value: FullName) => mixed,
}): React.Node => {
  return (
    <View style={styles.container}>
      <TitlePicker
        label="Title"
        value={name.title}
        onValueChange={val => onChange({...name, ['title']: val.toString()})}
      />
      <StringInput
        label="First name:"
        value={name.first}
        onChangeText={val => onChange({...name, ['first']: val})}
      />
      <StringInput
        label="Last name:"
        value={name.last}
        onChangeText={val => onChange({...name, ['last']: val})}
      />
    </View>
  );
};

export default FullNameInput;
