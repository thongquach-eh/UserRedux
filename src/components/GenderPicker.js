// @flow
import * as React from 'react';
import {StyleSheet, Text, View} from 'react-native';
import {Picker} from '@react-native-picker/picker';
import type {FieldProps} from 'redux-form';

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    flexWrap: 'wrap',
    alignItems: 'center',
  },
  label: {
    flexBasis: '25%',
    fontSize: 16,
    marginVertical: 3,
  },
  value: {
    flexBasis: '75%',
    fontSize: 16,
    marginVertical: 3,
  },
  pickerItem: {
    fontSize: 16,
  },
});

type GenderPickerProps = {
  ...FieldProps,
  label: string,
};

const GenderPicker = (props: GenderPickerProps): React.Node => {
  const {
    input: {value, onChange},
    label,
  } = props;

  return (
    <View style={styles.container}>
      <Text style={styles.label}>{label}</Text>
      <Picker
        style={styles.value}
        itemStyle={styles.pickerItem}
        selectedValue={value}
        onValueChange={onChange}>
        <Picker.Item label="male" value="male" />
        <Picker.Item label="female" value="female" />
      </Picker>
    </View>
  );
};

export default GenderPicker;
