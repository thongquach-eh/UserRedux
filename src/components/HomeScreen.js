// @flow
import * as React from 'react';
import {View, Text, Button, StyleSheet} from 'react-native';
import type {RouteProp} from '@react-navigation/native';
import type {StackNavigationProp} from '@react-navigation/stack';
import type {RootStackParamList} from '../types.js';

type HomeScreenProps = {
  navigation: StackNavigationProp<RootStackParamList, 'Home'>,
  route: RouteProp<RootStackParamList, 'Home'>,
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
  welcomeText: {
    fontSize: 50,
    textAlign: 'center',
  },
});

const HomeScreen = ({navigation, route}: HomeScreenProps): React.Node => {
  console.log(navigation);
  return (
    <View style={styles.container}>
      <Text style={styles.welcomeText}>Welcome to Mobile Training</Text>
      <Button
        title="Go to User List"
        onPress={() => navigation.navigate('UserList')}
      />
    </View>
  );
};

export default HomeScreen;
