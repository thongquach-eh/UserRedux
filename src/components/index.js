// @flow
export {default as HomeScreen} from './HomeScreen';
export {default as UserListScreen} from './UserListScreen';
export {default as UserDetailsScreen} from './UserDetailsScreen';
export {default as AddUserScreen} from './AddUserScreen';
export {default as EditUserScreen} from './EditUserScreen';
