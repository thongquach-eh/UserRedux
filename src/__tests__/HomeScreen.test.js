import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import {HomeScreen} from '../components';

describe('Home screen', () => {
  it('should render properly', () => {
    const component = <HomeScreen />;
    const tree = renderer.create(component).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
