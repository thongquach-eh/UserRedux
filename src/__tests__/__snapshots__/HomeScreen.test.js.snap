// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home screen should render properly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text
    style={
      Object {
        "fontSize": 50,
        "textAlign": "center",
      }
    }
  >
    Welcome to Mobile Training
  </Text>
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Go to User List
      </Text>
    </View>
  </View>
</View>
`;
